### ISSUE TITLE
dnscontrol push succeeds but subsequent dnscontrol preview out-of-date

### ISSUE THREAD

**Issue by StuartMorrisAU** [StackExchange/dnscontrol#2392]
Hi there,

I'm attempting to migrate a domain on Cloudflare to DNSControl. I ran:

```shell
$ dnscontrol get-zones --format=js --out=draft.js cloudflare - example.tld
```

I checked that over and copied it to `dnscontrol.js`, then ran:

```bash
$ dnscontrol check
No errors.
$ dnscontrol preview --domains example.tld
[INFO: Diff2 algorithm in use. Welcome to the future!]
******************** Domain: example.tld
1 correction (cloudflare)
#1: ± MODIFY example.tld MX (10 mx2.example.tld. ttl=1) -> (20 mx2.example.tld. ttl=1) id={id-string}
Done. 1 corrections.
```

That's fine as the priority should be 20 anyway. I `push`ed the change, then rerun `preview` and got:

```bash
$ dnscontrol preview --domains example.tld
[INFO: Diff2 algorithm in use. Welcome to the future!]
******************** Domain: example.tld
1 correction (cloudflare)
#1: ± MODIFY example.tld MX (10 mx2.example.tld. ttl=1) -> (20 mx2.example.tld. ttl=1) id={id-string}
Done. 1 corrections.
```

When I log into the Cloudflare dashboard, I can see the changes went through, so DNSControl isn't seeing the change for some reason. If I [get the zone](https://developers.cloudflare.com/api/operations/dns-records-for-a-zone-list-dns-records) using `curl`,  it returns the updated set of records.

The API key has rights to all zones and I've used it to migrate about ten domains on Cloudflare to DNSControl  so far without error. However, this is the largest zone (148 lines of DNS records in the config file).

Any ideas what might be going on here? I'm using:

dnscontrol "4.0.1" ("3cd907bfd3bc069b09b31e5d81ee6eb7790d2a9b") built 19 May 23 09:09 AEST

**Comment by tlimoncelli** [comment]
CC @hmoffatt and @tresni for suggestions

Hamish: Do you think this could be related to the bug you just fixed?

@StuartMorrisAU  Did the output of "get-zones" generate an MX() with preference 10 or 20?

Could you please try building DNSControl from the master branch and see if the bug exists there?

To build from master:

```
git clone https://github.com/StackExchange/dnscontrol.git
cd dnscontrol
git checkout master
go install
```
You will now find "dnscontrol" in your home dir's `bin` directory.

Tom

**Comment by StuartMorrisAU** [comment]
Thanks @tlimoncelli. Some MX records were priority 0 and others 10; however, they're assignment does seem a bit odd. There's two MX records per subdomain (10 subdomains plus the apex) and the priority should be 10 for the first and 20 for the second. `dnscontrol get-zones` set the priority randomly so that sometimes the first is 0 and other times 10. Similarly, the priority for the second MX record is sometimes 0 and other times 10. 

I get the following when I run `go install`:

```bash
build github.com/Azure/azure-sdk-for-go/sdk/azidentity: cannot load github.com/Azure/azure-sdk-for-go/sdk/azidentity: no Go source files
```

I didn't have go installed so I first ran `sudo apt install golang-go`.

Thaanks.

**Comment by hmoffatt** [comment]
> I didn't have go installed so I first ran `sudo apt install golang-go`.

If you're on Debian bullseye, the default golang is too old (1.15). Use bullseye-backports to get 1.19. I always install `golang` rather than `golang-go`, but the latter might be sufficient too.

**Comment by hmoffatt** [comment]
> Hamish: Do you think this could be related to the bug you just fixed?

I don't think so; that bug was only introduced in cloudflare-go v0.59, which we never used in dnscontrol. That bug caused MX priority updates to always fail.

I did some manual testing of changing MX priorities with master (with cloudflare-go v0.68) anyway and it worked fine.

I've put in a PR #2395 which allows you to log all the Cloudflare API interactions, so you can see what is being sent/received.

**Comment by StuartMorrisAU** [comment]
Backports was too old for focal so I downloaded 1.19 from the Go website. That let me build DNSControl without any errors.

`dnscontrol get-zones` now returns the proper priorities and I don't get the same issue if I `push` then `preview` like I was with 4.0.1.

Thanks for your help!

**Comment by StuartMorrisAU** [comment]
Confirmed fixed in 4.1.0.
